TickSpec/ServiceProvider.fs

```diff
@@ -36,11 +41,13 @@ type InstanceStore () =
         | Some elem -> Some(snd elem)
         | None -> None
 
+
     interface IDisposable with
         member __.Dispose() =
             instances
             |> Seq.map snd
             |> Seq.iter (function
+                | :? IAsyncDisposable as d -> disposeAsync(d)
```

Comment:
I have mixed feelings about introduction of DisposeAsync. On one hand a class should not have only DisposeAsync defined, but rather have both Dispose and DisposeAsync. Then the runtime itself decides what dispose method it calls, but only one of them is called in the end. On the other hand the DisposeAsync should not block and should not be used if it can't dispose the object asynchronously.

The way I suggest this to be implemented it is preferred to dispose objects asynchronously if they support it, but disposing synchronously the rest that can be disposed.

Supporting also async steps is a great idea, but probably much more complex to implement correctly. I'm open to suggestions.

We have a lot of cases where resources are asynchronously created and disposed with the following code:
``interface System.IDisposable with
        member this.Dispose() =
            (backgroundTask {
                // Do the disposal
            }).Wait()``

Reply:
You have mentioned the exact thing I don't like about this:
>  On one hand a class should not have only DisposeAsync defined, but rather have both Dispose and DisposeAsync. Then the runtime itself decides what dispose method it calls, but only one of them is called in the end.

So, what is the benefit of having the support here? If the instances have also `Dispose` then what is the benefit of calling `IAsyncDisposable` interface without propagating the awaitable task out of the scenario run?

On the other hand, it seems to me that it can cause issues when we start calling async methods from synchronous method in the style fire & forget. In lots of cases, it may mean that the disposal will not be finished as the main thread is blocked. Your workaround is more correct in the terms of running the disposal on background thread and synchronously waiting on the result on the main thread.

When we will go for another major version, I think that we are fine dropping the code specific for .NET Framework. While the code emit and breakpoints is a great functionality, as we are not able to get it running on netstandard, it is becoming irrelevant. Considering that, creating two variants of methods to run steps (synchronous and asynchronous) should not be overly complication. And the asynchronous variant would be able to invoke also async steps and IAsyncDisposable on the store while the synchronous variant would not support that.

Reply:
I agree that it makes no sense to add `DisposeAsync` support without supporting asynchronous steps. I think ideally the ServiceProvider should implement `IAsyncDisposable` instead of `IDispose` and itself be asynchronous.

Note that we had some time ago open pull request for async steps https://github.com/fsprojects/TickSpec/pull/23 but we closed it hoping for better implementation. 

Regarding emit code, I think it is still a miss that we don't support debugging feature files with .NET 6, I see that F# fsi is having the same issue but they are just working on a fix https://github.com/dotnet/fsharp/pull/12722, basically instead of using dynamic assemblies they use static, we can probably do the same, but I don't know how much effort it is and who and when can do that.

Reply:
Regarding emit, you are correct, it is a miss that we don't support it. However, I believe that we should not block all development just because it is harder due to old way we emit code.

It feels to me, that removing support for something that works only on .NET 4 and adding support for async would be a nice improvement. And we can work on bringing the support for breakpoints afterwards.

Yes, it would be awesome if we can keep the support for the things we have or even more awesome if we can make it happen even for .NET 6. We can also revisit what are all the code parts we do emit for, ideally we should not have the code duplicated. In other words, there are lots of things we can improve but we have to choose how to bring the biggest value.

Reply:
High level tangent/suggestion:

Ultimately someone looking at this repo in 6 months time is more interested in good .NET 8 support than some fantastic stuff in a video from 2009.

So, dont be afraid to consider:
1. make a `v2` branch which archives what's in the repo 
2. start a new world on `master`/`main` with only `net6.0` support with a prominent mention of the `v2` branch
2a. maybe remove old editions of test FWs examples etc
3. tag issues with 2.x vs 3.x milestones - i.e. any stuff about full fw / breakpoints etc can be 2.x
4. release stuff as `3.0.0-beta.1` etc until such time as you are finished cleaning things

Bottom line is that PRs like this (which is wonderful to see, welcome @iridin !) are much easier to do and/or likely to arrive if there is less ambiguous stuff that people need to traverse (I was going to suggest that the test FW parameterization stuff in this PR could go into a `Directory.build.props` but I think its better if everything is just rebooted to hardwired `net6.0` etc). I would suggest that e.g. explicit FSharp.Core references can also be stripped out. Things like switching from default FSharp.Core to explicit lower level reqs can then be reinstated as PRs.

(I did this in https://github.com/jet/equinox when moving from V2 to V3, and am about to do the same dance soon - I will cut a `v3` archive, do an aggressive tidy, and then start merging all the draft PRs I've been parking in there)

Please note this is just me talking - I'll scan stuff as a watcher but real decisions regarding this should be made by people who will have the time/interest to actually follow through.

Reply:
I fully agree.

We need to find way how to make good progress and starting by implementing first the code without IL and removing the emit path for now feels to me as the right step. It will make implementation of async steps and async dispose much easier.

Reply:
Sorry for taking so long to respond. I'm experimenting with making async tests all the way from testing framework into TickSpec steps. I will get back here once I have something that can be presented or discussed. Thanks